### PR TITLE
Add related identifier metadata fields

### PR DESCRIPTION
**Pull request by bess**
Work toward https://github.com/pulibrary/pdc_describe/issues/494

Co-authored-by: Hector Correa <hectorcorrea@users.noreply.github.com>

Note that this is not the full feature of related objects, but I've already been getting merge conflicts so I'd like to go ahead and get this into main before proceeding.

### CODE REVIEW

File: app/models/pdc_metadata/resource.rb [hunk 1]
```diff
@@ -111,6 +112,17 @@ def titles_from_json(resource, titles)
           end
         end
 
+        def related_objects_from_json(resource, related_objects)
+          return if related_objects.blank?
```
Comment (mccalluc) [inline app/models/pdc_metadata/resource.rb:116]: Is this needed? If `related_objects` is `[]`, then the `each` will be a no-op.

File: app/services/form_to_resource_service.rb [hunk 2]
```diff
@@ -76,6 +77,19 @@ def process_creators(params, resource)
         resource
       end
 
+      ## TODO: Do the right thing with blank form entries
```
Comment (mccalluc) [inline app/services/form_to_resource_service.rb:80]: How do we manage TODOs? Would it be good to file an issue and add the URL here?

File: app/views/works/_edit_javascript.html.erb [hunk 3]
```diff
@@ -81,6 +81,31 @@
       return html;
     }
 
+    // Related Objects
+    // related_identifier:, related_identifier_type:, relation_type
+    var addRelatedObjectHtml = function(num, related_identifier, related_identifier_type, relation_type) {
```
Comment (mccalluc) [inline app/views/works/_edit_javascript.html.erb:86]:
```suggestion
    var addRelatedObjectHtml = function({num, related_identifier, related_identifier_type, relation_type}) {
```
This is getting long enough that I'd worry about always getting the parameters in the correct order: Using parameter destructuring might be a little safer.

Comment (mccalluc) [reply app/views/works/_edit_javascript.html.erb:86]: or supply default values?
```suggestion
    var addRelatedObjectHtml = function({num, related_identifier='', related_identifier_type='', relation_type=''}) {
```

File: app/views/works/_edit_javascript.html.erb [hunk 4]
```diff
@@ -354,6 +385,21 @@
...
+    if ($(".related-object-data").length ==0) {
+      // Add an empty related object for the user to fill it out
+      var num = incrementCounter("#related_object_count");
+      addRelatedObjectHtml(num, "", "", "");
+    } else {
+      // Add existing related objects for editing
+      var related_objects = $(".related-object-data");
+      for(i = 0; i < related_objects.length; i++) {
+        var related_object = related_objects[i];
+        addRelatedObjectHtml(related_object.dataset.num, related_object.dataset.relatedIdentifier, related_object.dataset.relatedIdentifierType, related_object.dataset.relationType);
```
Comment (mccalluc) [inline app/views/works/_edit_javascript.html.erb:399]:
```suggestion
        const {num, relatedIdentifier, relatedIdentifierType, relationType} = related_object.dataset
        addRelatedObjectHtml(num, relatedIdentifier, relatedIdentifierType, relationType);
```

File: app/views/works/_edit_javascript.html.erb [hunk 5]
```diff
@@ -354,6 +385,21 @@
...
+      }
+    }
+
     if ($(".contributor-data").length ==0) {
```
Comment (mccalluc) [inline app/views/works/_edit_javascript.html.erb:403]:
```suggestion
    if ($(".contributor-data").length == 0) {
```

File: app/views/works/show.html.erb [hunk 6]
```diff
@@ -103,6 +103,14 @@
     </p>
   <% end %>
 
+  <% if @work.resource.related_objects.count > 0 %>
+    <p>Related Objects:
+      <% @work.resource.related_objects.each_with_index do |related_object, ix| %>
```
Comment (mccalluc) [inline app/views/works/show.html.erb:108]:
```suggestion
      <% @work.resource.related_objects.each do |related_object| %>
```

File: spec/models/work_spec.rb [hunk 7]
```diff
@@ -79,6 +79,15 @@
     end
   end
 
+  context "with related objects" do
+    subject(:work) { FactoryBot.create(:distinct_cytoskeletal_proteins_work) }
+    it "has related objects" do
+      expect(work.resource.related_objects.first.related_identifier).to eq "https://www.biorxiv.org/content/10.1101/545517v1"
+      expect(work.resource.related_objects.first.related_identifier_type).to eq "arXiv"
+      expect(work.resource.related_objects.first.relation_type).to eq "IsCitedBy"
```
Comment (mccalluc) [inline spec/models/work_spec.rb:87]: below I see `isCitedBy`... is it case insensitive, or is something not being tested? or we capitalize it for display on the page?

File: spec/system/cytoskeletal_form_submission_spec.rb [hunk 8]
```diff
@@ -71,13 +74,20 @@
...
       click_on "v-pills-curator-controlled-tab"
       fill_in "doi", with: doi
       fill_in "ark", with: ark
       click_on "Create"
       expect(page).to have_content "marked as draft"
       cytoskeletal_work = Work.last
       expect(cytoskeletal_work.title).to eq title
+      expect(cytoskeletal_work.resource.related_objects.first.related_identifier).to eq related_identifier
+      expect(cytoskeletal_work.resource.related_objects.first.related_identifier_type).to eq related_identifier_type
+      expect(cytoskeletal_work.resource.related_objects.first.relation_type).to eq relation_type
```
Comment (mccalluc) [inline spec/system/cytoskeletal_form_submission_spec.rb:90]: Might assign `cytoskeletal_work.resource.related_objects.first` to a variable, mostly just to keep the line length down.

**Review by mccalluc — DISMISSED** [review]
A couple comments on JS you can take or leave. The one point that confused me is the capitalization of `IsCitedBy`.

File: app/services/form_to_resource_service.rb (hunk 2 above)
Comment (bess) [reply app/services/form_to_resource_service.rb:80]: This is a very short-term TODO. I'm not done with this feature yet, so I'm just reminding myself that I'm not done.

File: app/views/works/_edit_javascript.html.erb (hunk 4 above)
Comment (bess) [reply app/views/works/_edit_javascript.html.erb:399]: Sorry, my js isn't the greatest. I don't understand what is being suggested here.

File: spec/models/work_spec.rb (hunk 7 above)
Comment (bess) [reply spec/models/work_spec.rb:87]: We're going to be working with a controlled vocabulary and I don't actually know which capitalization option is correct. I was going to tackle that next.

File: app/views/works/_edit_javascript.html.erb (hunk 3 above)
Comment (mccalluc) [reply app/views/works/_edit_javascript.html.erb:86]: (Might change the JS syntax in the future, but this is fine for now.)

File: app/views/works/_edit_javascript.html.erb [hunk 9]
```diff
@@ -354,7 +385,23 @@
       }
     }
 
-    if ($(".contributor-data").length ==0) {
+    // Load any existing related objects into the edit form. 
+    // If there are any related objects they should appear in hidden <span> tags.
+    if ($(".related-object-data").length ==0) {
```
Comment (mccalluc) [inline app/views/works/_edit_javascript.html.erb:390]:
```suggestion
    if ($(".related-object-data").length == 0) {
```

File: app/views/works/_edit_javascript.html.erb [hunk 10]
```diff
@@ -354,7 +385,23 @@
...
+      // Add an empty related object for the user to fill it out
+      var num = incrementCounter("#related_object_count");
+      addRelatedObjectHtml(num, "", "", "");
+    } else {
+      // Add existing related objects for editing
+      var related_objects = $(".related-object-data");
+      for(i = 0; i < related_objects.length; i++) {
+        var related_object = related_objects[i];
```
Comment (mccalluc) [inline app/views/works/_edit_javascript.html.erb:398]:
```suggestion
      for (const related_object of $(".related-object-data")) {
```

**Review by mccalluc — DISMISSED** [review]

**Review by mccalluc — APPROVED** [review]

**Pull request merged**